### ISSUE TITLE
switch to pytorch aten breaks WITH_CUDA=0 build

### ISSUE THREAD

**Issue by skimo-openhub** [facebookresearch/TensorComprehensions#334]
Since c8a44248d980a8d7c88b929024f795409e3d3020, TC no longer builds with
`WITH_CUDA=0`.
This is the error message I get:

```
cd /home/skimo/git/c2isl/third-party/pytorch/aten/build/src/ATen/test && /usr/bin/cmake -E cmake_link_script CMakeFiles/cudnn_test.dir/link.txt --verbose=1
/usr/bin/g++-6   -DUSE_AVX2 -msse3 -DUSE_SSE3 --std=c++11 -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-unused-parameter -Wno-unknown-pragmas -Wno-vla -fexceptions  -D_FORCE_INLINES -D_MWAITXINTRIN_H_INCLUDED -D__STRICT_ANSI__ -fopenmp -O3    CMakeFiles/cudnn_test.dir/cudnn_test.cpp.o  -o cudnn_test -Wl,-rpath,/home/skimo/git/c2isl/third-party/pytorch/aten/build/src/ATen -rdynamic ../libATen.so ../cpu/tbb/libtbb_static.a -lpthread -ldl -lblas -llapack -lrt -lm ../cpu/cpuinfo/libcpuinfo.a 
CMakeFiles/cudnn_test.dir/cudnn_test.cpp.o: In function `____C_A_T_C_H____T_E_S_T____0()':
cudnn_test.cpp:(.text+0x275c5): undefined reference to `at::native::getCudnnHandle()'
cudnn_test.cpp:(.text+0x27bb7): undefined reference to `cudnnDropoutGetStatesSize'
cudnn_test.cpp:(.text+0x27c18): undefined reference to `cudnnCreateDropoutDescriptor'
cudnn_test.cpp:(.text+0x27c42): undefined reference to `cudnnDestroyDropoutDescriptor'
```

TC builds fine without `WITH_CUDA=0` and it used to build fine with `WITH_CUDA=0`
before that commit.

**Comment by ftynse** [comment]
Apparently, Aten still looks for Cudnn even if Cuda is disabled.  I get this in the output

```
disabling CUDA because NO_CUDA is set                                           
-- Found cuDNN: v7.0.5  (include: /usr/local/cuda/include, library: /usr/local/cuda/lib/libcudnn.so)
-- CuDNN not found. Compiling without CuDNN support
```

and `CUDNN_FOUND`is set to `TRUE`.

**Comment by ftynse** [comment]
This fixed the problem for me

```
diff --git a/aten/CMakeLists.txt b/aten/CMakeLists.txt
index 9648d1907..e248058c6 100644
--- a/aten/CMakeLists.txt
+++ b/aten/CMakeLists.txt
@@ -468,7 +468,9 @@ else()
   add_subdirectory(src/THCS)
 endif()

-find_package(CuDNN)
+IF (NOT NO_CUDA)
+  find_package(CuDNN)
+ENDIF()
 IF(NOT AT_CUDA_ENABLED OR NOT CUDNN_FOUND)
   MESSAGE(STATUS "CuDNN not found. Compiling without CuDNN support")
   set(AT_CUDNN_ENABLED 0)
```

@prigoyal @zdevito should I send this patch to ATen?

**Comment by skimo-openhub** [comment]
On Tue, Apr 24, 2018 at 02:02:29AM -0700, ftynse wrote:
> This fixed the problem for me
> 
> ```
> diff --git a/aten/CMakeLists.txt b/aten/CMakeLists.txt
> index 9648d1907..e248058c6 100644
> --- a/aten/CMakeLists.txt
> +++ b/aten/CMakeLists.txt
> @@ -468,7 +468,9 @@ else()
>    add_subdirectory(src/THCS)
>  endif()
> 
> -find_package(CuDNN)
> +IF (NOT NO_CUDA)
> +  find_package(CuDNN)
> +ENDIF()
>  IF(NOT AT_CUDA_ENABLED OR NOT CUDNN_FOUND)
>    MESSAGE(STATUS "CuDNN not found. Compiling without CuDNN support")
>    set(AT_CUDNN_ENABLED 0)
> ```

Thanks.  This seems to work for me too.

skimo

**Comment by prigoyal** [comment]
please feel free to send it to pytorch repo.

**Comment by ftynse** [comment]
https://github.com/pytorch/pytorch/pull/6905

**Comment by prigoyal** [comment]
PR is merged.

**Comment by ftynse** [comment]
I think we need to bump ATen in TC before closing this.

**Comment by ftynse** [comment]
there seems to be another bug in cmake `list sub-command REMOVE_ITEM requires two or more arguments.` if I bump it :/

**Comment by prigoyal** [comment]
what commit hash did you try for ATen? I can take a look and try to help :)

**Comment by ftynse** [comment]
`e7babb1890b1658886835ff4627d8b00aa84bae1`

this should do the trick

```
diff --git a/aten/src/ATen/cpu/tbb/CMakeLists.txt b/aten/src/ATen/cpu/tbb/CMakeLists.txt
index 9c8d04405..6891c8eb7 100644
--- a/aten/src/ATen/cpu/tbb/CMakeLists.txt
+++ b/aten/src/ATen/cpu/tbb/CMakeLists.txt
@@ -50,7 +50,7 @@ endif()
 file(GLOB tbb_src "${TBB_ROOT_DIR}/src/tbb/*.cpp" "${TBB_ROOT_DIR}/src/old/*.cpp")
 list(APPEND tbb_src ${TBB_ROOT_DIR}/src/rml/client/rml_tbb.cpp)
 file(GLOB to_remove "${TBB_ROOT_DIR}/src/old/test*.cpp")
-list(REMOVE_ITEM tbb_src ${to_remove})
+list(REMOVE_ITEM tbb_src "${to_remove}")

 set(tbbmalloc_static_src
   src/tbbmalloc/backend.cpp
```

**Comment by skimo-openhub** [comment]
Original problem fixed by #440 because it ends up updating pytorch.
@ftynse, please open a separate issue for your other problem.

**Comment by ftynse** [comment]
does not seem to be a problem for me anymore
